src/util/metadata/side_metadata/helpers.rs

```diff
@@ -25,6 +25,32 @@ pub(super) fn address_to_contiguous_meta_address(
     }
 }
 
+/// Performs reverse address translation from contiguous metadata bits to data addresses.
+///
+/// Arguments:
+/// * `metadata_spec`: The side metadata spec. It should be contiguous side metadata.
+/// * `metadata_addr`; The metadata address. Returned by [`address_to_contiguous_meta_address`].
+/// * `shift`: The bit shift for the metadata. Returned by [`meta_byte_lshift`].
+pub(super) fn contiguous_meta_address_to_address(
+    metadata_spec: &SideMetadataSpec,
+    metadata_addr: Address,
+    shift: u8,
+) -> Address {
+    let rshift = (LOG_BITS_IN_BYTE as i32) - metadata_spec.log_num_of_bits as i32;
+    let relative_meta_addr = metadata_addr - metadata_spec.get_absolute_offset();
+
+    let data_addr_intermediate = if rshift >= 0 {
+        relative_meta_addr << rshift
+    } else {
+        relative_meta_addr >> (-rshift)
+    };
+
+    let data_addr = (data_addr_intermediate << metadata_spec.log_bytes_in_region)
+        + ((shift as usize) << (metadata_spec.log_bytes_in_region - metadata_spec.log_num_of_bits));
```

Comment:
It is unlikely, but in theory `metadata_spec.log_bytes_in_region - metadata_spec.log_num_of_bits` may underflow.

If `log_num_of_bits >= 3` (i.e. `rshift <= 0`), then `shift` must be zero.  We can split this statement into the `if` and `else` branch above.  When `rshift >= 0`, we only use `metadata_spec.log_bytes_in_region`; when `rshift < 0`, we take `metadata_spec.log_num_of_bits` into consideration.

Reply:
Done. I added a few more tests to cover this case.